QuickEditor: ProfileCard using forceRefresh option for Avatar

hamorillo opened the pull request:
Closes #281 

### Description

This PR uses a different approach than #340 to solve #281.

As discussed [here](https://github.com/Automattic/Gravatar-SDK-Android/pull/340#discussion_r1784288363) we can implement a `forceRefresh` param in the `Avatar` composable.

With this approach, we don't need a new Avatar composable. In addition, we increase the versatility of our current component.

If `forceRefresh` is enabled, everytime the `Avatar` composable is recomposed we'll update the cache buster internally. When we switch off that flag, we'll continue using the last `cacheBuster` used.

On the QE side, as the profile does not change, we need to force the recomposition. In this case, we use the "key" with the selected image ID.

**Important:** To me, this approach looks better than #340 and also reduces the number of composables we need to maintain. However, the avatar is not always updated as expected because we request the new avatar too fast. We'll need to introduce a delay to postpone the recomposition a bit. You can reproduce it by switching between already uploaded avatars. Wdyt?

<details>
<summary> Without delay </summary>

https://github.com/user-attachments/assets/d299f9fb-2dee-41f1-9313-61e93d4ba529
</details>

<details>
<summary> With Delay </summary>

https://github.com/user-attachments/assets/0c948716-90e4-4c68-bd6b-2142164ed40e
</details>

### Testing Steps

1. Smoke test the `Avatar Tab` in the demo app. Everything should work as before.
2. In the `Avatar Update Tab` with the Network Inspector opened, switch between different images. 
3. Verify the request contains the `size` param and the `cacheBuster`

<img width="673" alt="image" src="https://github.com/user-attachments/assets/5f6cc826-eba5-47b1-9155-dda188a67439">

hamorillo committed:
Avatar composable with forceRefresh option

When the avatar is updated, we need to refresh the current image. For that purpose, we need the cache buster in the request.

On the composable side, as the state does not change (because the avatar URL is the same), we need to force the recomposition. In this case, we use the "key" with the selected image ID.

AdamGrzybkowski commented on gravatar-quickeditor/src/main/java/com/gravatar/quickeditor/ui/avatarpicker/AvatarPicker.kt:
```diff
@@ -146,10 +147,13 @@ internal fun AvatarPicker(uiState: AvatarPickerUiState, onEvent: (AvatarPickerEv
-                profile = uiState.profile,
-                modifier = Modifier.padding(horizontal = 16.dp),
-            )
+            key(uiState.emailAvatars?.selectedAvatarId) {
```
This one is new for me. What exactly is its usage here? Is it to only recompose when `selectedAvatarId` is changed? 

Won't that prevent updates when the `uiState.profile` is modified?

hamorillo replied:
> Is it to only recompose when selectedAvatarId is changed?

Yes. We know when the avatar is updated, but the Composable, which shows the avatar, doesn't depend on that state (the profile will be the same). So, using the key will force its recomposition based on the `selectedAvatarId` changes.

> Won't that prevent updates when the uiState.profile is modified

As far as I know, `ProfileCard` should still be updated if `uiState.profile` is modified.

AdamGrzybkowski commented on gravatar-ui/src/main/java/com/gravatar/ui/components/atomic/Avatar.kt:
```diff
@@ -27,15 +31,23 @@ import com.gravatar.ui.skeletonEffect
+    forceRefresh: Boolean = false,
 ) {
     val sizePx = with(LocalDensity.current) { size.roundToPx() }
+    var cacheBuster by rememberSaveable { mutableStateOf<String?>(null) }
+
+    if (forceRefresh) {
+        cacheBuster = System.currentTimeMillis().toString()
+    }
+
```
When `forceRefresh` is true the `cacheBuster` will be updated on each recomposition. What's the advantage of using `rememberSaveable` for it?

hamorillo replied:
At somepoint `forceRefresh` can be switched to `false` and, for example, rotate the device. The idea is to still use the previous cacheBuster.

AdamGrzybkowski reviewed (commented):
The current version is without the delay, right? Where would that delay be implemented?

hamorillo reviewed (commented):
> The current version is without the delay, right? 

Yes.

> Where would that delay be implemented?

That's a good question. I don't like the delay approach, but I couldn't come up with a better solution. I was thinking of adding the deley in the viewmodel, something like:

```kotlin
is Result.Success -> {
        delay(X ms)
        _uiState.update { currentState ->
```

But as I'm not a fan of this, I wanted to check with you if you have a better idea.

AdamGrzybkowski commented:
>However, the avatar is not always updated as expected because we request the new avatar too fast. We'll need to introduce a delay to postpone the recomposition a bit. You can reproduce it by switching between already uploaded avatars. Wdyt?

@pinarol Have you had a similar problem with the cache buster param not being enough to show the latest selected image? How did you handle that on iOS?

pinarol commented:
> @pinarol Have you had a similar problem with the cache buster param not being enough to show the latest selected image? How did you handle that on iOS?

When the user selects a different avatar, we directly use the recently selected image's URL to show it, so we don't use cache busting here. But this is also news to me if cache busting wasn't enough. 🤔

hamorillo committed:
Adding a delay until inform that the Avatar has been switched

In spite of using the cacheBuster, the backend needs some time until it returns the right avatar after the switch.

hamorillo commented:
I've added the delay before notifying the avatar switch.

I've also created #354 to remove the URL from the callback.

hamorillo commented on gravatar-quickeditor/src/main/java/com/gravatar/quickeditor/ui/avatarpicker/AvatarPickerViewModel.kt:
```diff
@@ -34,6 +34,10 @@ internal class AvatarPickerViewModel(
     private val avatarRepository: AvatarRepository,
     private val fileUtils: FileUtils,
 ) : ViewModel() {
+    private companion object {
```
With less than 800 ms, I still see frequent issues.

hamorillo commented on gravatar-quickeditor/src/main/java/com/gravatar/quickeditor/ui/avatarpicker/AvatarPickerViewModel.kt:
```diff
@@ -34,6 +34,10 @@ internal class AvatarPickerViewModel(
     private val fileUtils: FileUtils,
 ) : ViewModel() {
+    private companion object {
+        const val AVATAR_SWITCH_DELAY = 800L
```
With less than 800 ms, I still see frequent issues.

pinarol replied:
Thanks for testing 👍 
cc: @etoledom
